### PR TITLE
feat: add gitea remote driver.

### PR DESCRIPTION
**Pull request by appleboy**
Since Gitea already supports status API. https://github.com/go-gitea/gitea/pull/1332 It is time to create a new remote driver for Gitea.

cc @tboerger @lunny @bkcsoft

ref: https://github.com/drone/drone/issues/1978

### CODE REVIEW

File: remote/gitea/gitea_test.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	"testing"
+
+	"github.com/drone/drone/model"
+	"github.com/drone/drone/remote/gitea/fixtures"
+
+	"github.com/franela/goblin"
+	"github.com/gin-gonic/gin"
+)
+
+func Test_gogs(t *testing.T) {
```
Comment (lunny) [inline remote/gitea/gitea_test.go]: Gitea

File: remote/gitea/gitea.go [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+	}
+	return toPerm(repo.Permissions), nil
+}
+
+// File fetches the file from the Gitea repository and returns its contents.
+func (c *client) File(u *model.User, r *model.Repo, b *model.Build, f string) ([]byte, error) {
+	client := c.newClientToken(u.Token)
+	ref := b.Commit
+
+	// TODO gogs does not yet return a sha with the pull request
```
Comment (lunny) [inline remote/gitea/gitea.go]: gitea

File: remote/gitea/gitea.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/drone/drone/model"
+	"github.com/drone/drone/remote"
+
+	"github.com/go-gitea/go-sdk/gitea"
```
Comment (tboerger) [inline remote/gitea/gitea.go]: code.gitea.io/sdk should be the correct import path.

File: remote/gitea/gitea.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+	Machine     string
+	Username    string
+	Password    string
+	PrivateMode bool
+	SkipVerify  bool
+}
+
+const (
+	DescPending  = "the build is pending"
+	DescRunning  = "the buils is running"
```
Comment (tboerger) [inline remote/gitea/gitea.go]: Typo

File: remote/gitea/fixtures/handler.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+func createRepoHook(c *gin.Context) {
+	in := struct {
+		Type string `json:"type"`
+		Conf struct {
+			Type string `json:"content_type"`
+			URL  string `json:"url"`
+		} `json:"config"`
+	}{}
+	c.BindJSON(&in)
+	if in.Type != "gogs" ||
```
Comment (lunny) [inline remote/gitea/fixtures/handler.go]: gitea

File: remote/gitea/gitea.go [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+		Username:    opts.Username,
+		Password:    opts.Password,
+		PrivateMode: opts.PrivateMode,
+		SkipVerify:  opts.SkipVerify,
+	}, nil
+}
+
+// Login authenticates an account with Gitea using basic authenticaiton. The
+// Gitea account details are returned when the user is successfully authenticated.
+func (c *client) Login(res http.ResponseWriter, req *http.Request) (*model.User, error) {
```
Comment (tboerger) [inline remote/gitea/gitea.go]: OAuth2 provider is not done yet, right?

File: remote/gitea/fixtures/hooks.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
+package fixtures
+
+// Sample Gogs push hook
```
Comment (lunny) [inline remote/gitea/fixtures/hooks.go]: Gitea

File: remote/gitea/gitea.go (hunk 2 above)
Comment (tboerger) [inline remote/gitea/gitea.go]: Is this still the case for gitea?

**Comment by appleboy** [comment]
@tboerger @lunny Fix testing first.

File: remote/gitea/fixtures/handler.go [hunk 8]
```diff
@@ -79,7 +79,6 @@ const repoPayload = `
 {
   "owner": {
     "login": "test_name",
-    "username": "test_name",
```
Comment (appleboy) [inline remote/gitea/fixtures/handler.go]: See https://developer.github.com/v3/users/#get-a-single-user

File: remote/gitea/gitea.go (hunk 6 above)
Comment (appleboy) [inline remote/gitea/gitea.go]: Yes. you are right.

File: drone/server/server.go [hunk 9]
```diff
@@ -218,6 +218,37 @@ var Command = cli.Command{
 			Usage:  "gogs skip ssl verification",
```
Comment (lunny) [inline drone/server/server.go]: gitea

File: remote/gitea/fixtures/handler.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
+package fixtures
+
+import (
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+)
+
+// Handler returns an http.Handler that is capable of handling a variety of mock
+// Bitbucket requests and returning mock responses.
```
Comment (mrueg) [inline remote/gitea/fixtures/handler.go]: gitea

Comment (appleboy) [inline remote/gitea/fixtures/handler.go]: https://github.com/drone/drone/pull/2017/commits/d631966a913807723461718fb1730d652bfc3c6d done

File: remote/gitea/gitea.go [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+		return gitea.StatusPending
+	case model.StatusRunning:
+		return gitea.StatusPending
+	case model.StatusSuccess:
+		return gitea.StatusSuccess
+	case model.StatusFailure, model.StatusError:
+		return gitea.StatusFailure
+	case model.StatusKilled:
+		return gitea.StatusFailure
+	default:
```
Comment (bkcsoft) [inline remote/gitea/gitea.go]: Though I'm not sure what `StatusDeclined` is actually :)

File: remote/gitea/helper.go [hunk 12]
```diff
@@ -186,7 +186,7 @@ func parsePullRequest(r io.Reader) (*pullRequestHook, error) {
 }
 
 // fixMalformedAvatar is a helper function that fixes an avatar url if malformed
-// (currently a known bug with gogs)
+// (currently a known bug with gitea)
```
Comment (bkcsoft) [inline remote/gitea/helper.go]: Link to bug issue?

File: drone/server/server.go [hunk 13]
```diff
@@ -218,6 +218,37 @@ var Command = cli.Command{
...
 		cli.BoolFlag{
+			EnvVar: "DRONE_GITEA",
+			Name:   "gitea",
+			Usage:  "gitea driver is enabled",
+		},
+		cli.StringFlag{
+			EnvVar: "DRONE_GITEA_URL",
+			Name:   "gitea-server",
+			Usage:  "gitea server address",
+			Value:  "https://github.com",
```
Comment (lunny) [inline drone/server/server.go]: ^

File: remote/gitea/gitea.go [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/drone/drone/model"
+	"github.com/drone/drone/remote"
+
```
Comment (tboerger) [inline remote/gitea/gitea.go]: We are not on Gitea, drop this empty line :)

File: remote/gitea/gitea.go (hunk 11 above)
Comment (tboerger) [inline remote/gitea/gitea.go]: This is related to the gating of drone if the gating have been declined. So I think gitea.StatusWarning makes sense.

File: remote/gitea/gitea.go [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+		Avatar: expandAvatar(c.URL, account.AvatarURL),
+	}, nil
+}
+
+// Auth is not supported by the Gitea driver.
+func (c *client) Auth(token, secret string) (string, error) {
+	return "", fmt.Errorf("Not Implemented")
+}
+
+// Teams is not supported by the Gitea driver.
```
Comment (tboerger) [inline remote/gitea/gitea.go]: I think this comment is outdated.

File: remote/gitea/gitea.go [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+	}
+
+	var teams []*model.Team
+	for _, org := range orgs {
+		teams = append(teams, toTeam(org, c.URL))
+	}
+	return teams, nil
+}
+
+// TeamPerm is not supported by the Gitea driver.
```
Comment (tboerger) [inline remote/gitea/gitea.go]: Is this really not supported?

File: remote/gitea/gitea.go [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+	}
+	return toPerm(repo.Permissions), nil
+}
+
+// File fetches the file from the Gitea repository and returns its contents.
+func (c *client) File(u *model.User, r *model.Repo, b *model.Build, f string) ([]byte, error) {
+	client := c.newClientToken(u.Token)
+	ref := b.Commit
+
+	// TODO gitea does not yet return a sha with the pull request
```
Comment (tboerger) [inline remote/gitea/gitea.go]: Is this still valid for Gitea?

File: remote/gitea/gitea.go [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+	cfg, err := client.GetFile(r.Owner, r.Name, ref, f)
+	return cfg, err
+}
+
+// FileRef fetches the file from the Gitea repository and returns its contents.
+func (c *client) FileRef(u *model.User, r *model.Repo, ref, f string) ([]byte, error) {
+	return c.newClientToken(u.Token).GetFile(r.Owner, r.Name, ref, f)
+}
+
+// Status is not supported by the Gitea driver.
```
Comment (tboerger) [inline remote/gitea/gitea.go]: Yep, outdated comment as well :D

File: remote/gitea/gitea.go [hunk 19]
```diff
@@ -0,0 +1,344 @@
...
+		Events: []string{"push", "create", "pull_request"},
+		Active: true,
+	}
+
+	client := c.newClientToken(u.Token)
+	_, err := client.CreateRepoHook(r.Owner, r.Name, hook)
+	return err
+}
+
+// Deactivate is not supported by the Gitea driver.
```
Comment (tboerger) [inline remote/gitea/gitea.go]: Please take a look at my pending pull-request, I have started to integrate that for Gogs some time ago.

File: remote/gitea/helper.go [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func parsePullRequest(r io.Reader) (*pullRequestHook, error) {
+	pr := new(pullRequestHook)
+	err := json.NewDecoder(r).Decode(pr)
+	return pr, err
+}
+
+// fixMalformedAvatar is a helper function that fixes an avatar url if malformed
+// (currently a known bug with gitea)
```
Comment (tboerger) [inline remote/gitea/helper.go]: Again, is that still the case for Gitea?

File: remote/gitea/parse.go [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/drone/drone/model"
+)
+
+const (
+	hookEvent       = "X-Gogs-Event"
```
Comment (tboerger) [inline remote/gitea/parse.go]: X-Gitea-Event? AFAIK we got this as fallback.

File: remote/gitea/gitea.go (hunk 19 above)
Comment (tboerger) [inline remote/gitea/gitea.go]: For reference: https://github.com/drone/drone/pull/1752

File: remote/gitea/parse.go (hunk 21 above)
Comment (appleboy) [inline remote/gitea/parse.go]: We can create next PR to improvement this event.

Comment (appleboy) [inline remote/gitea/parse.go]: because we need to update Gitea source code first.

File: remote/gitea/gitea.go (hunk 19 above)
Comment (appleboy) [inline remote/gitea/gitea.go]: Waiting for reference PR #1752

File: remote/gitea/gitea.go (hunk 11 above)
Comment (appleboy) [inline remote/gitea/gitea.go]: Add `model.StatusDeclined`

File: remote/gitea/gitea.go (hunk 17 above)
Comment (bkcsoft) [inline remote/gitea/gitea.go]: No, we send a sha now :)

File: remote/gitea/gitea.go [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+	DescRunning  = "the build is running"
+	DescSuccess  = "the build was successful"
+	DescFailure  = "the build failed"
+	DescCanceled = "the build canceled"
+	DescBlocked  = "the build is pending approval"
+	DescDeclined = "the build was rejected"
+)
+
+// getStatus is a helper functin that converts a Drone
+// status to a GitHub status.
```
Comment (mrueg) [inline remote/gitea/gitea.go]: s/GitHub/gitea/

File: remote/gitea/parse.go [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+	actionOpen = "opened"
+	actionSync = "synchronize"
+
+	stateOpen = "open"
+
+	refBranch = "branch"
+	refTag    = "tag"
+)
+
+// parseHook parses a Bitbucket hook from an http.Request request and returns
```
Comment (mrueg) [inline remote/gitea/parse.go]: s/Bitbucket/Gitea/

File: remote/gitea/gitea.go [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+	if c.Password != "" {
+		return &model.Netrc{
+			Login:    c.Username,
+			Password: c.Password,
+			Machine:  c.Machine,
+		}, nil
+	}
+	return &model.Netrc{
+		Login:    u.Token,
+		Password: "x-oauth-basic",
```
Comment (bkcsoft) [inline remote/gitea/gitea.go]: `"x-oauth-basic"` ?

File: remote/gitea/gitea_test.go [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+				g.Assert(repos[0].Name).Equal(fakeRepo.Name)
+				g.Assert(repos[0].FullName).Equal(fakeRepo.Owner + "/" + fakeRepo.Name)
+			})
+			g.It("Should handle a not found error", func() {
+				_, err := c.Repos(fakeUserNoRepos)
+				g.Assert(err != nil).IsTrue()
+			})
+		})
+
+		g.It("Should register repositroy hooks", func() {
```
Comment (bkcsoft) [inline remote/gitea/gitea_test.go]: `"repositroy" => "repository"`

File: remote/gitea/gitea_test.go [hunk 26]
```diff
@@ -0,0 +1,194 @@
...
+			g.Assert(string(raw)).Equal("{ platform: linux/amd64 }")
+		})
+
+		g.It("Should return a repository file from a ref", func() {
+			raw, err := c.File(fakeUser, fakeRepo, fakeBuildWithRef, ".drone.yml")
+			g.Assert(err == nil).IsTrue()
+			g.Assert(string(raw)).Equal("{ platform: linux/amd64 }")
+		})
+
+		g.It("Should return nil frome send build status", func() {
```
Comment (bkcsoft) [inline remote/gitea/gitea_test.go]: `"frome" => "from"`

File: remote/gitea/gitea.go [hunk 27]
```diff
@@ -0,0 +1,345 @@
...
+	DescPending  = "the build is pending"
+	DescRunning  = "the build is running"
+	DescSuccess  = "the build was successful"
+	DescFailure  = "the build failed"
+	DescCanceled = "the build canceled"
+	DescBlocked  = "the build is pending approval"
+	DescDeclined = "the build was rejected"
+)
+
+// getStatus is a helper functin that converts a Drone
```
Comment (xxxtonixxx) [inline remote/gitea/gitea.go]: typo in `functin`

Comment (appleboy) [inline remote/gitea/gitea.go]: fix

File: remote/gitea/parse.go (hunk 21 above)
Comment (tboerger) [inline remote/gitea/parse.go]: You are right, I thought we already added Gitea there: https://github.com/go-gitea/gitea/blob/master/models/webhook.go#L539-L542

Comment (appleboy) [inline remote/gitea/parse.go]: I think we can improve this with another PR. This PR just support commit status and create the new Gitea driver.

File: remote/gitea/gitea.go (hunk 19 above)
Comment (tboerger) [inline remote/gitea/gitea.go]: The reference PR is only for the Gogs driver, not for Gitea.

Comment (appleboy) [inline remote/gitea/gitea.go]: @tboerger Yes. I can create another PR to implement `Deactivate`.

**Comment by vtolstov** [comment]
Very nice, when this can be merged ?=)

File: vendor/vendor.json [hunk 28]
```diff
@@ -7,6 +7,12 @@
 			"revision": ""
 		},
 		{
+			"checksumSHA1": "Vrt1uoeOOk9fp1aa9d7huQhzexU=",
+			"path": "code.gitea.io/sdk/gitea",
```
Comment (bradrydzewski) [inline vendor/vendor.json]: should this be `github.com/go-gitea/go-sdk`? 

when I visit `code.gitea.io/sdk/gitea` I don't see any source code (maybe I'm not understanding the UI though). I would not feel comfortable with a dependency that developers cannot trace back to the origin source.

Comment (tboerger) [inline vendor/vendor.json]: Nope, this is the correct import path :). It's simple go get magic:

```
# curl -s https://code.gitea.io/sdk/?go-get=1 | grep code.gitea.io/sdk
		<meta name="go-import" content="code.gitea.io/sdk git https://github.com/go-gitea/go-sdk">
		<meta name="go-source" content="code.gitea.io/sdk https://github.com/go-gitea/go-sdk https://github.com/go-gitea/go-sdk/tree/master{/dir} https://github.com/go-gitea/go-sdk/blob/master{/dir}/{file}#L{line}">
```

Comment (tboerger) [inline vendor/vendor.json]: Just to give you more background information: We introduced a custom import path directly at the beginning to be able to switch to dog-fooding at some point without distroying previous installs.

Comment (lunny) [inline vendor/vendor.json]: ^

File: remote/gitea/gitea.go (hunk 19 above)
Comment (bradrydzewski) [inline remote/gitea/gitea.go]: yes, we should implement Deactivate in this PR. Without the ability to deactivate a repository cannot be re-enabled correctly

File: remote/gitea/gitea.go (hunk 16 above)
Comment (bradrydzewski) [inline remote/gitea/gitea.go]: this function is no longer used so we can remove from Remote

File: remote/gitea/gitea.go (hunk 17 above)
Comment (bradrydzewski) [inline remote/gitea/gitea.go]: Should we update this to include sha?

File: remote/gitea/gitea.go (hunk 19 above)
Comment (bradrydzewski) [inline remote/gitea/gitea.go]: yes, we should implement Deactivate as part of this PR. Without the ability to deactivate, a repository cannot be re-enabled correctly

**Comment by appleboy** [comment]
@bradrydzewski @lunny @bkcsoft @tboerger @DblK

https://github.com/drone/drone/pull/2017/commits/7641092070c4588eb3c80bb74d82e3b3e84645b7

I already finished the implements of `deactivate` and refactor the `file` method finally. Please help to review again. Please make sure you need to update some changes from https://github.com/go-gitea/gitea/pull/1755 before testing this PR.

**Comment by strk** [comment]
which release of Drone will first have this support ?

**Comment by appleboy** [comment]
@strk I guess the 1.2 version.

**Comment by strk** [comment]
I mean Drone version - latest is 0.8

**Comment by appleboy** [comment]
@strk 0.7 version

**Pull request merged**